ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -898,16 +922,24 @@ static bool ShouldHandleFileOpEvent(
     PerfTracer* perfTracer,
     vfs_context_t context,
     const vnode_t vnode,
-    kauth_action_t action,
+    const char* fromPath, // used only for hard-link & rename events, nullptr otherwise
 
     // Out params:
-    VirtualizationRootHandle* root,
+    VirtualizationRootHandle* vnodeRoot,
+    VirtualizationRootHandle* fromProvider, // when fromPath is non-null, the handle of the provider where the file was originally located is written to here
     FsidInode* vnodeFsidInode,
     int* pid)
 {
+    bool pidIsValid = false;
```

Comment:
pidIsSet?  I take it pidIsValid means we've made the pid call and want to avoid doing it twice